### ISSUE TITLE
Blaze renders documents in incorrect order

### ISSUE THREAD

**Issue by arggh** [meteor/blaze#231]
I tried implementing a drag & drop sorting feature with [Dragula](https://github.com/bevacqua/dragula) and Blaze.

I wanted to use Dragula for the drag and drop UI all the way until the `drop`-event, then let Blaze and Meteor handle the actual ordering and re-rendering of the list.

As instructed in one of the [comments](https://github.com/bevacqua/dragula/issues/188#issuecomment-148102910) in Dragula's issue https://github.com/bevacqua/dragula/issues/188 ,

I created my `drop` event handler in Dragula to prevent DOM manipulation, like this:

```js
Template.body.onRendered(function() {
  // Get the container element for Dragula
  const container = this.$('.documents').get(0);

  // Initialize Dragula
  const drake = dragula([container]);

  // Attach handler for drop event to prevent DOM manipulation and trigger moving
  drake.on('drop', (el, target, source, sibling) => {
    // Retrieve new order for the dragged element
    const newIndex = Array.from(el.parentNode.children).indexOf(el);

    // Get dragged document id
    const id = el.getAttribute('data-id');

    // Cancel the drag&drop, preventing DOM manipulation by Dragula
    drake.cancel(true);

    // Call Meteor method to change dragged Document's order
    Meteor.call('moveTo', id, newIndex);
  });
});
```

I list these documents, sorted by order like so:

```js
Template.body.helpers({
  /**
   * Get all Documents, sorted by order
   */
  docs() {
    return Documents.find({}, {
      sort: {
        order: 1
      }
    });
  }
});
```

```handlebars
{{#each docs}}
    <div class="document">
        <span class="order">{{this.order}}</span> {{this.name}}
    </div>
{{/each}}
```

**The problem is**, if I drag the _second_ item to _first_ place (order changing from `1` to `0`), the order is changed in Mongo like it should, and even the `order` property that is printed in the element changes accordingly, **but the order of elements in DOM remains in the previous state.**

<img width="856" alt="blaze-dragula" src="https://cloud.githubusercontent.com/assets/17210302/22072553/66a15086-ddab-11e6-966a-57399d52d603.png">

This only seems to happen when dragging an element to a new order that is exactly one smaller than it's current order, so for example `3` -> `2`. The same thing happened with the other library I tried before Dragula, [Sortable](https://github.com/RubaXa/Sortable).

If I call the same method without dragging the element and move it from index `1` to `0`, Blaze moves the DOM elements to the correct order like it should.

I don't know much about Blaze's internals, but Dragula (Sortable too) is a hugely popular drag and drop library and it would be really nice to be able to use it with Blaze.

Does anybody have ideas what Dragula is doing to cause the incorrect order of elements, or why Blaze really doesn't play well in this case? @mitar ? 

### Repro
https://github.com/arggh/blaze-sort-issue

1. meteor npm install
2. meteor
3. click "Add document" 3-4 times
4. Try to drag second element to first place and pay attention what happens (wrong order!)
5. Refresh
6. Try to do the same re-arrangement by clicking the other button

**Comment by arggh** [comment]
Actually, I was too hasty in my conclusions. The incorrect rendering of items happens every time you move an item up on the list by one step, it doesn't have to be the second item.

**Comment by mitar** [comment]
I confirm this is happening. I do not have time to investigate at the moment why. :-(

Can you try maybe with an older version of Blaze? 2.1.9? Just to make sure we did not introduce this lately. Because a very similar thing I am using in my code with jQuery sortable. So not sure why this would not work.

**Comment by arggh** [comment]
I can confirm it's happening also with `2.1.9`.

**Comment by mitar** [comment]
Hm, [this is how I am using jQuery sortable](https://github.com/peer/mind/blob/master/packages/peermind/meeting/display.coffee#L97) and it works for me in 99% cases (there are some when it breaks, but it is very rare).

**Comment by arggh** [comment]
If I do these steps:

1. remove the line of code that updates the order property of documents on drop
2. start the app
3. make note of the DOM in fresh state
4. drag the second element in list to first place and let Dragula cancel the drop
5. I make sure DOM is in exact same state as in 2.
6. click the second button in the reproduction to update the order property without dragging
7. the same thing happens, sorting fails and the element with order 1 is first while order 0 is second.

So, initiating the drag does somehow confuse Blaze. It's not about a race condition or the DOM being in a funny state while re-rendering, as it's exactly the same.

As potential clues, what the Dragula library is doing while dragging:

- Creates a clone of the dragged element, placing it outside of the container
- Adds a `gu-transit`class to the original element
- Moves the original element inside the container element in respect to dragging, to visualize the potential drop placement between other elements

**Comment by mitar** [comment]
I think you should read #186 and whole discussion in https://github.com/meteor/meteor/issues/2003. This is a really long issue in Blaze. For jQuery there [was a patch](https://github.com/meteor/meteor/issues/2003#issuecomment-263044202) which was not accepted upstream. Maybe you can get something similar to be accepted in Dragula.

One workaround I heard is that you store away the original content, duplicate whole DOM for the list, allow library for dragging to do its job, and then at dropping you replace it back with original DOM (as made by Blaze) and then let Blaze to update it. Ugly.

**Comment by arggh** [comment]
Thanks for digging those issues up.

I worked around it already by adding a cursor observer for the query of those draggable items, and if any document has it's `order` property changed (so, after dragging and updating DB), I hide the list with an `#if` block and show it again after flush. This puts the elements in correct order.

**Comment by jeffm24** [comment]
I realize that this thread is pretty old, but I noticed something that someone might be able to make sense of.  From what I can tell, the problem seems to stem from the ObserveSequence movedTo callback which is processed within the Blaze.Each function.  I noticed that when you do a move like @arggh was describing (i.e. [1, 2, 3] -> [1, 3, 2]), it always seems to say that item 3 was moved from index 2 to 1 even if you actually moved item 2 from index 1 to 2.  Consequently, it calls eachView._domrange.moveMember with the fromIndex and toIndex swapped.  I'm not entirely sure about the internals past this point, but could this be causing Blaze to get confused about where items are currently positioned in the DOM @mitar ?  I am using SortableJS for clarity, but I have seen the same behavior happen with both that and dragula.
